### PR TITLE
[export] dynamic_shapes serialization, load/dump

### PR DESCRIPTION
**Pull request by pianpwk**
Adds utility functions `_dump_dynamic_shapes` and `_load_dynamic_shapes`. 

- `_dump_dynamic_shapes`: dynamic shapes spec -> serialized format:
    - takes in the `dynamic_shapes` pytree object you'd feed into `export()`, and dumps into serialized format
- `_load_dynamic_shapes`: serialized format -> dynamic shapes spec
    - takes the serialized format, and produces a `dynamic_shapes` object you feed into `export()`

For example with dumping:
```
dx = Dim("dx", min=4, max=16)
dy = dx + 1

inputs = (
    [
        torch.randn(4, 4),
        torch.randn(5, 4),
    ],
    torch.randn(4),
    torch.randn(4, 4),
    "hello",
)
dynamic_shapes = {
    "a": [
        (dx, 4),
        (dy, 4),
    ],
    "b": (Dim.AUTO,),
    "c": None,
    "d": None,
}
out = _dump_dynamic_shapes(dynamic_shapes, inputs)
```

would generate the following output:
```
DynamicShapesSpec(
    dynamic_shapes=(
        [
            ['dx', 4],
            ['dx + 1', 4],
        ],
        ['_DimHint.STATIC'],
        ['_DimHint.STATIC', '_DimHint.STATIC'],
        None,
    ),
    dims={
        'dx': RootDim(
            min=4,
            max=16,
            derived=['dx + 1'],
        ),
    },
)
```

The serialized format contains 2 keys, `dynamic_shapes` and `dims.` 
- `dynamic_shapes` is the pytree structure matching the input to `export()`, with strings in place of Dim names and enums, and ints/Nones otherwise. Each tensor is represented with a list of shapes, non-tensors with Nones.
- `dims` contain min/max range and derived dims info for each root dim.

The test cases show some roundtrippability guarantees for these functions. Definitely taking naming suggestions for them :)

Follow up: utility function to extract serializable format from ExportedProgram.

### CODE REVIEW

File: test/export/test_export.py [hunk 1]
```diff
@@ -293,19 +293,55 @@ def _test_export_same_as_eager(self, f, args, kwargs=None):
...
+                    model,
+                    inputs,
+                    dynamic_shapes=_loads(_dumps(_specs, inputs)),
+                )
+                ep_extract = export(
+                    model,
+                    inputs,
+                    dynamic_shapes=_loads(_extracts(ep)),
+                )
+                eps.extend([ep_serdes, ep_extract])
```
Comment (pianpwk) [inline test/export/test_export.py:335]: these 3 ExportedPrograms should all have the same shape-related behavior

File: test/export/test_export.py [hunk 2]
```diff
@@ -6897,6 +6933,7 @@ def forward(self, x, y, z):
                 ((4, 4), (4, 4), (4, 3)),
                 ((4, 4), (5, 4), (4, 5)),
             ],
+            test_serdes=True,
```
Comment (pianpwk) [inline test/export/test_export.py:6936]: patch serdes testing for auto dynamic shapes

File: test/export/test_export.py [hunk 3]
```diff
@@ -7039,8 +7082,116 @@ def forward(self, x, y):
...
+                ([(24, 4), (25, 4)], (12,), (4, 4), "hello"),
+            ],
+            [
+                ([(16, 4), (17, 4)], (8,), (5, 5), "hello"),
+            ],
+            test_serdes=True,
+        )
+        self.assertExpectedInline(
+            _dumps(dynamic_shapes, inputs),
+            """{'dynamic_shapes': ([['2*dz', 4], ['2*dz + 1', 4]], ['dz'], [None, None], None), 'dims': {'dz': {'min': 4, 'max': 16, 'derived': ['2*dz', '2*dz + 1']}}}"""
```
Comment (pianpwk) [inline test/export/test_export.py:7136]: yeahh so `_extracts` goes off of the symbol names. We could improve this and go off placeholder names.

File: test/export/test_export.py [hunk 4]
```diff
@@ -7039,8 +7082,116 @@ def forward(self, x, y):
...
+        extracted = _extracts(export(Foo(), args, kwargs, dynamic_shapes=dynamic_shapes))
+        self.assertExpectedInline(
+            extracted,
+            """{'dynamic_shapes': (['s0', 4], [['s2', 8], ['s2', 8]], ['s2 + 1', 8]), 'dims': {'s0': {'min': 2, 'max': None, 'derived': []}, 's2': {'min': 2, 'max': None, 'derived': ['s2 + 1']}}}""",
+        )
+        # deserialize & export with both specs, and check that the extracted specs match up, with some roundtrippability
+        extracted_from_dumped = _extracts(export(Foo(), args, kwargs, dynamic_shapes=_loads(dumped)))
+        extracted_from_extracted = _extracts(export(Foo(), args, kwargs, dynamic_shapes=_loads(extracted)))
+        self.assertEqual(extracted, extracted_from_dumped)
+        self.assertEqual(extracted, extracted_from_extracted)
```
Comment (pianpwk) [inline test/export/test_export.py:7193]: roundtrippability guarantees

File: torch/export/dynamic_shapes.py [hunk 5]
```diff
@@ -1191,3 +1193,280 @@ def apply_fixes(path, dim, dummy):
...
+
+
+def _postprocess_serialized_shapes(dynamic_shapes, dims):
+    """
+    Sorts dims and returns dictionary format
+    """
+    from torch.utils._sympy.numbers import int_oo
+
+    dims = {
+        k: {
```
Comment (angelayi) [inline torch/export/dynamic_shapes.py:1205]: nit: maybe we can define this as a dataclass above, and then use dataclasses.asdict to dump it into a dictionary?

File: torch/export/dynamic_shapes.py [hunk 6]
```diff
@@ -1191,3 +1193,280 @@ def apply_fixes(path, dim, dummy):
...
+            "min": v["min"],
+            "max": None if v["max"] is int_oo else v["max"],
+            "derived": sorted(v["derived"]),
+        }
+        for k, v in sorted(dims.items())
+    }
+    return {
+        "dynamic_shapes": dynamic_shapes,
+        "dims": dims,
+    }
```
Comment (angelayi) [inline torch/export/dynamic_shapes.py:1215]: same as above, to convert this to a dataclass

File: torch/export/dynamic_shapes.py [hunk 7]
```diff
@@ -1191,3 +1193,280 @@ def apply_fixes(path, dim, dummy):
         return dim  # unchanged dim
 
     return _tree_map_with_path(apply_fixes, dynamic_shapes, dynamic_shapes)
+
```
Comment (angelayi) [inline torch/export/dynamic_shapes.py:1196]: nit: what if we moved these functions to _export/serde? 

also suggestion for the names, can we suffix the functions with `_dynamic_shapes`? Like, `dump_dynamic_shapes`

File: torch/export/dynamic_shapes.py [hunk 8]
```diff
@@ -1191,3 +1193,280 @@ def apply_fixes(path, dim, dummy):
...
+        }
+        for k, v in sorted(dims.items())
+    }
+    return {
+        "dynamic_shapes": dynamic_shapes,
+        "dims": dims,
+    }
+
+
+def _dumps(dynamic_shapes, args, kwargs=None):
```
Comment (angelayi) [inline torch/export/dynamic_shapes.py:1218]: type annotations please~~

File: torch/export/dynamic_shapes.py [hunk 9]
```diff
@@ -1191,3 +1193,280 @@ def apply_fixes(path, dim, dummy):
...
+            return _DimHint.AUTO
+        elif val == "_DimHint.STATIC":
+            return _DimHint.STATIC
+        assert isinstance(val, str)
+        return dim_cache[val]
+
+    return tree_map(deserialize_shape, dynamic_shapes)
+
+
+def _extracts(ep: ExportedProgram):
```
Comment (angelayi) [inline torch/export/dynamic_shapes.py:1385]: Would it be more composable to have a function which get the normal dynamic_shapes dictionary? And then users can call dump/load on that directly?

File: test/export/test_export.py [hunk 10]
```diff
@@ -7039,8 +7082,118 @@ def forward(self, x, y):
...
+            ],
+            test_serdes=True,
+        )
+        self.assertExpectedInline(
+            _dumps(dynamic_shapes, inputs),
+            """{'dynamic_shapes': ([['2*dz', 4], ['2*dz + 1', 4]], ['dz'], [None, None], None), 'dims': {'dz': {'min': 4, 'max': 16, 'derived': ['2*dz', '2*dz + 1']}}}"""
+        )
+        self.assertExpectedInline(
+            _extracts(ep),
+            """{'dynamic_shapes': ([['2*d0', 4], ['2*d0 + 1', 4]], ['d0'], [4, 4], None), 'dims': {'d0': {'min': 4, 'max': 16, 'derived': ['2*d0', '2*d0 + 1']}}}""",
```
Comment (pianpwk) [inline test/export/test_export.py:7140]: `_extracts()` just enumerates the symbols, doing this as a placeholder if anyone has better ideas. Could use the placeholder names + indices.

File: torch/export/dynamic_shapes.py (hunk 9 above)
Comment (pianpwk) [reply torch/export/dynamic_shapes.py:1385]: Oh, users can call `_loads(_extracts(ep))`, I guess it's just the extra step

File: torch/export/dynamic_shapes.py [hunk 11]
```diff
@@ -1191,3 +1193,283 @@ def apply_fixes(path, dim, dummy):
...
+    would generate the following output:
+    ```
+    {
+        'dynamic_shapes': (
+            [
+                ['dx', 4],
+                ['dx + 1', 4],
+            ],
+            ['Dim.AUTO'],
+            [None, None],
```
Comment (avikchaudhuri) [inline torch/export/dynamic_shapes.py:1265]: This should be `Dim.STATIC`. I'm working on a PR that will begin warning on `None` in dim positions.

File: torch/export/dynamic_shapes.py [hunk 12]
```diff
@@ -1191,3 +1193,283 @@ def apply_fixes(path, dim, dummy):
...
+            None,
+        ),
+        'dims': {
+            'dx': {
+                'min': 4,
+                'max': 16,
+                'derived': ['dx + 1'],
```
Comment (avikchaudhuri) [inline torch/export/dynamic_shapes.py:1272]: I'm not sure I follow, are you saying we'll have a list of everything derived from `dx` here?

File: torch/export/dynamic_shapes.py [hunk 13]
```diff
@@ -1191,3 +1193,283 @@ def apply_fixes(path, dim, dummy):
...
+            },
+            'dy': {
```
Comment (avikchaudhuri) [inline torch/export/dynamic_shapes.py:1274]: This should be better than just having dummy values.

I'm assuming you considered and rejected having the definition `dx + 1` in here? Why does that make things easier / harder?

File: torch/export/dynamic_shapes.py (hunk 9 above)
Comment (avikchaudhuri) [reply torch/export/dynamic_shapes.py:1385]: I'd prefer `_extracts` to be in a separate PR, and probably as a method on `ExportedProgram` that has the behavior @angelayi suggested.

Comment (avikchaudhuri) [reply torch/export/dynamic_shapes.py:1385]: Also (didn't read below yet), how do you keep the names `dx` etc. around, do you instead use symbols like `s0`?

File: torch/export/dynamic_shapes.py (hunk 13 above)
Comment (pianpwk) [reply torch/export/dynamic_shapes.py:1274]: It's because when we construct DerivedDims, we're just applying a function on the root Dim, and min/max ranges aren't used - they're just derived from the root's properties.

File: torch/export/dynamic_shapes.py (hunk 9 above)
Comment (pianpwk) [reply torch/export/dynamic_shapes.py:1385]: ah the names disappear, right now I just enumerate them into `d0, d1, ...` (using the symbols seem to have some strict/non-strict mismatches). I'll split this into a separate PR.

**Review by avikchaudhuri — APPROVED** [review]
Much better, thanks for making the requested changes!
